keanu-project/src/main/java/io/improbable/keanu/network/LambdaSection.java

```diff
@@ -0,0 +1,121 @@
+ * all of the vertices that affects it (upstream), stopping at observed or probabilistic vertices.
+ * <p>
+ * For example:
+ * <p>
+ * A = SomeDistribution(...)
+ * B = A.cos()
+ * C = SomeDistribution(B, ...)
+ * D = C.times(2)
+ * <p>
+ * The downstream Lambda Section of A would be [A, B, C]
+ * The upstream Lambda Section of D would be [D, C]
+ * The upstream Lambda Section of C would be [C, B, A]
+ */
+@Value
+public class LambdaSection {
+
+    private static final Predicate<Vertex> ADD_ALL = vertex -> true;
+    private static final Predicate<Vertex> PROBABILISTIC_AND_OBSERVED_ONLY = vertex -> vertex.isObserved() || vertex.isProbabilistic();
+
+    private final Set<Vertex> allVertices;
+    private final Set<Vertex> latentAndObservedVertices;
+
+    private LambdaSection(Set<Vertex> allVertices) {
+        this.allVertices = allVertices;
+        this.latentAndObservedVertices = allVertices.stream()
+            .filter(PROBABILISTIC_AND_OBSERVED_ONLY)
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * @param aVertex                 the starting vertex
+     * @param includeNonProbabilistic false if only the probabilistic and observed are wanted
+     * @return All upstream vertices up to probabilistic or observed vertices if includeNonProbabilistic
+     * is true. All upstream probabilistic or observed vertices stopping at probabilistic or observed if
+     * includeNonProbabilistic is false.
+     */
+    public static LambdaSection getUpstreamLambdaSection(Vertex<?> aVertex, boolean includeNonProbabilistic) {
+
+        Predicate<Vertex> shouldAdd = includeNonProbabilistic ? ADD_ALL : PROBABILISTIC_AND_OBSERVED_ONLY;
+
+        Set<Vertex> upstreamVertices = getVerticesDepthFirst(
+            aVertex,
+            Vertex::getParents,
+            shouldAdd
+        );
+
+        return new LambdaSection(upstreamVertices);
+    }
+
+    /**
+     * @param aVertex                 the starting vertex
+     * @param includeNonProbabilistic false if only the probabilistic and observed are wanted
+     * @return All downstream vertices up to probabilistic or observed vertices if includeNonProbabilistic
+     * is true. All downstream probabilistic or observed vertices stopping at probabilistic or observed if
+     * includeNonProbabilistic is false.
+     */
+    public static LambdaSection getDownstreamLambdaSection(Vertex<?> aVertex, boolean includeNonProbabilistic) {
+
+        Predicate<Vertex> shouldAdd = includeNonProbabilistic ? ADD_ALL : PROBABILISTIC_AND_OBSERVED_ONLY;
+
+        Set<Vertex> downstreamVertices = getVerticesDepthFirst(
+            aVertex,
+            Vertex::getChildren,
+            shouldAdd
+        );
+
+        return new LambdaSection(downstreamVertices);
+    }
+
+    /**
+     * @param vertex       Vertex to start propagation from
+     * @param nextVertices The next vertices to move to give a current vertex
```

Comment:
Not sure what this means?

Reply:
It's a lambda that either gets childern or parents. I've updated the doc to show these as examples.